perf: reduce DB calls to `GetWorkspaceByAgentID` via caching workspace info

cstyan opened the pull request:
This query is still relatively expensive because of its call volume, at least 1 million times per day over the last week. After https://github.com/coder/coder/pull/20430 the two main remaining call paths are via the Workspace Agents Stats and Metadata APIs.

Disclaimer: Blink helped write tests, identify usage of `database.Workspace` fields in downstream functions, identify existing pattern for and write the ticker function, and write comments.

This PR adds caching of the Workspace information on the AgentAPI struct for usage in both child APIs, Notably, for the Metadata API we need to also implement a "fast path" for the section that currently calls `GetWorkspaceByAgentID` since it's nested within a dbauthz call. We do this by attaching the relevant RBAC object from the Workspace object to the request context, then checking in `UpdateWorkspaceAgentMetadata` if we have the RBAC object in the context _and_ if it is correct for authorization purposes. If we do not have the object or it is invalid we fall back to still calling `GetWorkspaceByAgentID`.

Note that while the workspace agent API struct is constructed at startup and (IIUC) is persistent for the lifetime of the agent/workspace, and makes a call to `GetWorkspaceByAgentID` during this process, we also need to handle updating of some of the cached fields since the entries within the database _can_ change as part of the workspace prebuild process. We could potentially due this via subscribing, via pubsub, to updates about prebuild workspace claims. However, for now this has simply been solved with a 5m ticker to make a call to `GetWorkspaceByAgentID`.

This should mean we on average call `GetWorkspaceByAgentID` once every 5 minutes per agent, as opposed to multiple times per-minute per agent for the Stats API and again for the Metadata API. Even in the worst case scenario of a workspace prebuild being claimed at t=N and the next ticker not occurring until t=N+5minutes we should have at most 5 minutes of the current behaviour for each agent in a workspace.

**NOTE:** there is a remaining issue here; caching of the workspace for stats as it is currently implemented is actually invalid, at least for the ~5 minute period between when a prebuild is claimed and the next update of the ticker the cached Owner information will be incorrect and we'll update some incorrect metrics + skip the entire `if !workspace.IsPrebuild()` block.

We can either:
1. revert this optimization here and still just always call `GetWorkspaceByAgentID` for every stats update
2. similar to option 1, but only _always_ call `GetWorkspaceByAgentID` if the workspace is a prebuild, and the first time it _is not_ we could grab the lock and update the cached workspace information
3. do the _right_ thing for both stats and metadata, and hook into pubsub to get the prebuild claim update ASAP while still avoiding extra calls to `GetWorkspaceByAgentID`

cstyan committed:
remove unnecessary calls to GetWorkspaceAgentByID for stats api calls

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
modify workspace caching, significantly reduce calls to
GetWorkspaceByAgentID for agent metadata updates

Signed-off-by: Callum Styan <callumstyan@gmail.com>

Emyrk commented on coderd/agentapi/metadata.go:
```diff
@@ -17,10 +17,11 @@ import (
-	Pubsub   pubsub.Pubsub
-	Log      slog.Logger
+	AgentFn       func(context.Context) (database.WorkspaceAgent, error)
+	RBACContextFn func(context.Context) context.Context
```
Is there a reason why we need to pass this function in? I know it is on the MetadataAPI. I would just prefer a code structure without having to pass a function like this since it is only used in 1 place.

Maybe just pass the `CachedWorkspaceFields` type to both the API and the metadata api?

Emyrk commented on coderd/agentapi/metadata.go:
```diff
@@ -107,6 +108,9 @@ func (a *MetadataAPI) BatchUpdateMetadata(ctx context.Context, req *agentproto.B
+	// Inject RBAC object into context for dbauthz fast path, avoid having to
+	// call GetWorkspaceByAgentID on every metadata update.
+	ctx = a.RBACContextFn(ctx)
 	err = a.Database.UpdateWorkspaceAgentMetadata(ctx, dbUpdate)
```
In case `ctx` is used in the function later, we should require the cache to be populated per request.
```suggestion
	err = a.Database.UpdateWorkspaceAgentMetadata(a.RBACContextFn(ctx), dbUpdate)
```

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -5507,6 +5524,30 @@ func (q *querier) UpdateWorkspaceAgentLogOverflowByID(ctx context.Context, arg d
+				return ErrNoActor
+			}
+			err := q.auth.Authorize(ctx, act, policy.ActionUpdate, rbacObj)
+			if err == nil {
```
Add a comment that failures will always skip the cache and hit the database in case the cache is stale.

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -5507,6 +5524,30 @@ func (q *querier) UpdateWorkspaceAgentLogOverflowByID(ctx context.Context, arg d
+	// claim, there may be up to a 5-minute window where this uses stale owner_id.
+	if rbacObj, ok := WorkspaceRBACFromContext(ctx); ok {
+		// Verify the RBAC object is valid (has required fields).
+		if rbacObj.Owner != "" && rbacObj.OrgID != "" && rbacObj.Owner != uuid.Nil.String() && rbacObj.OrgID != uuid.Nil.String() {
```
We need to add a check the type is `ResourceWorkspace`

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -201,6 +201,23 @@ func ActorFromContext(ctx context.Context) (rbac.Subject, bool) {
+// authorization data for the duration of an agent connection.
+func WithWorkspaceRBAC(ctx context.Context, rbacObj rbac.Object) context.Context {
+	return context.WithValue(ctx, workspaceRBACContextKey{}, rbacObj)
+}
```
Add some guards to make sure the `rbacObj` type is `Workspace`, that ACL's are empty, that owners are set.

cstyan committed:
Add an IsEmpty check for rbac object.

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
don't need to wait on ctx after tickerfunc, plus log when we've exited
and invalidate the cached workspace

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
only wrap context for the one database call

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
formatting

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
guard against the rbac object being invalid when adding it to the
context

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
lint

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
fix tests never finishing; we need to wait on the tickerFunc finishing
but also not log at error level

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
don't cache prebuilds

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
use the opts workspaceID in refreshCachedWorkspace now that we don't
auto cache in the API constructor if the workspace is a prebuild

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
these shouldn't have been removed

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
lint and fmt + move workspace rbac context stuff to a separate file

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
store pointer to cached workspace instead of functions

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
we need to cache the AutostartSchedule for the stats api

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
move cached workspace struct to its own file, include mutex on the
struct, add clear + update functions and very basic tests

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
make fmt and linter happy

Signed-off-by: Callum Styan <callumstyan@gmail.com>

Emyrk commented on coderd/agentapi/api.go:
```diff
@@ -114,6 +119,24 @@ func New(opts Options) *API {
+				String: workspace.AutostartSchedule.String,
+				Valid:  workspace.AutostartSchedule.Valid,
+			},
+		}
```
Given there is a mutex in the CachedWorkspaceFields, we should probably always assign to a pointer. 

In the `api` struct that is. So we need an `else` condition to init an empty Cache

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -5507,6 +5507,32 @@ func (q *querier) UpdateWorkspaceAgentLogOverflowByID(ctx context.Context, arg d
+	// claim, there may be up to a 5-minute window where this uses stale owner_id.
+	if rbacObj, ok := WorkspaceRBACFromContext(ctx); ok {
+		// Verify the RBAC object is valid (has required fields).
+		if !rbacObj.IsEmpty() && rbacObj.Type == rbac.ResourceWorkspace.Type {
```
`WorkspaceRBACFromContext` should never return an invalid `RBACObject` imo. The `WithWorkspaceRBAC` protects against these errors like the wrong type.

So we can trust it and not guard against things if `ok == true`

cstyan committed:
make cached workspace a pointer on the API struct

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
clear the cached workspace on DB error

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
address more review feedback; attempt to simplify cached workspace is
valid/present checks, plus ensure proper fallback (less noisy logging)
when a cached workspace is not found/valid in dbauthz

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
remove IsEmpty from rbac object, we'd need to handle too many rbac
object types there

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan commented on coderd/agentapi/metadata.go:
```diff
@@ -107,7 +109,15 @@ func (a *MetadataAPI) BatchUpdateMetadata(ctx context.Context, req *agentproto.B
+		// Don't use error level here; that will fail the call but in reality an invalid object here is okay
+		// since we will then just fallback to the actual GetWorkspaceByAgentID call in dbauthz.
+		a.Log.Warn(ctx, "cached RBAC Workspace context wrapping was invalid", slog.Error(err))
+	}
```
this now only logs if the workspace _was_ cached and the rbac object was still somehow invalid

cstyan committed:
typo in function name

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
refactor so that functions which will now make use of the cached
workspace info. can't misuse database.Workspace and access fields which
in theory may not be set

Signed-off-by: Callum Styan <callumstyan@gmail.com>

Emyrk reviewed (approved):
The `WorkspaceIdentity` does add the type safety. :+1:

cstyan committed:
fmt and lint again

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
cleanup/refactor tests, we no longer cache prebuilds but still want to
test that fields are updated via the 5m update

Signed-off-by: Callum Styan <callumstyan@gmail.com>

cstyan committed:
reduce logging for expected paths

Signed-off-by: Callum Styan <callumstyan@gmail.com>
